pkg/snet/directtp/client.go

```diff
@@ -0,0 +1,468 @@
+	LocalAddr       string
+	Table           pktable.PKTable
+	AddressResolver arclient.APIClient
+}
+
+type client struct {
+	conf           Config
+	mu             sync.Mutex
+	done           chan struct{}
+	once           sync.Once
+	log            *logging.Logger
+	porter         *porter.Porter
+	packetFilter   *pfilter.PacketFilter
+	packetListener net.PacketConn
+	visorsConn     net.PacketConn
+	listener       net.Listener
+	listeners      map[uint16]*tplistener.Listener // key: lPort
+}
+
+// NewClient creates a net Client.
+func NewClient(conf Config) Client {
+	return &client{
+		conf:      conf,
+		log:       logging.MustGetLogger(conf.Type),
+		porter:    porter.New(porter.MinEphemeral),
+		listeners: make(map[uint16]*tplistener.Listener),
+		done:      make(chan struct{}),
+	}
+}
+
+// Serve serves the listening portion of the client.
+func (c *client) Serve() error {
+	switch c.conf.Type {
+	case tptypes.STCP, tptypes.STCPR:
+		if c.listener != nil {
+			return ErrAlreadyListening
+		}
+	case tptypes.SUDPH:
+		if c.packetListener != nil {
+			return ErrAlreadyListening
+		}
+	}
+
+	l, err := c.getListener()(c.conf.LocalAddr)
+	if err != nil {
+		return err
+	}
+
+	c.listener = l
+
+	if c.conf.Type == tptypes.STCPR {
+		_, port, err := net.SplitHostPort(c.listener.Addr().String())
+		if err != nil {
+			return err
+		}
+
+		if err := c.conf.AddressResolver.BindSTCPR(context.Background(), c.conf.Type, port); err != nil {
+			return fmt.Errorf("bind %v: %w", c.conf.Type, err)
+		}
+	}
+
+	go func() {
+		c.log.Infof("listening on addr: %v", c.listener.Addr())
+
+		for {
+			if err := c.acceptConn(); err != nil {
+				c.log.Warnf("failed to accept incoming connection: %v", err)
+
+				if !tphandshake.IsHandshakeError(err) {
+					c.log.Warnf("stopped serving")
+					return
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (c *client) acceptConn() error {
+	if c.isClosed() {
+		return io.ErrClosedPipe
+	}
+
+	conn, err := c.listener.Accept()
+	if err != nil {
+		return err
+	}
+
+	remoteAddr := conn.RemoteAddr()
+
+	c.log.Infof("Accepted connection from %v", remoteAddr)
+
+	var lis *tplistener.Listener
+
+	hs := tphandshake.ResponderHandshake(func(f2 tphandshake.Frame2) error {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		var ok bool
+		if lis, ok = c.listeners[f2.DstAddr.Port]; !ok {
+			return errors.New("not listening on given port")
+		}
+
+		return nil
+	})
+
+	connConfig := tpconn.Config{
+		Log:       c.log,
+		Conn:      conn,
+		LocalPK:   c.conf.PK,
+		LocalSK:   c.conf.SK,
+		Deadline:  time.Now().Add(tphandshake.Timeout),
+		Handshake: hs,
+		FreePort:  nil,
+		Encrypt:   true,
+		Initiator: false,
+	}
+
+	wrappedConn, err := tpconn.NewConn(connConfig)
+	if err != nil {
+		return fmt.Errorf("newConn: %w", err)
+	}
+
+	if err := lis.Introduce(wrappedConn); err != nil {
+		return fmt.Errorf("introduce: %w", err)
+	}
+
+	return nil
+}
+
+// Dial dials a new sudp.Conn to specified remote public key and port.
```

Comment:
why is it `sudp.Conn`? sorry, might be missing something

Reply:
You're right. It shouldn't be `sudp.Conn`.